viewing_party/party.py

```diff
@@ -1,23 +1,122 @@
+        return movie
+    else:
+        return None
+    
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
 
+def watch_movie(user_data, title):
+
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+    return user_data
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    # an empty watched list is 0.0
+    if not user_data["watched"]:
+        return 0.0
+    # caculate total rating
+    total_rating = 0
+    for movie in user_data["watched"]:
+        total_rating += movie["rating"]
+
+    avg_rating = total_rating / len(user_data["watched"])
 
+    return avg_rating
 
+def get_most_watched_genre(user_data):
+    # an empty list return None
+    if not user_data["watched"]:
+        return None
+    
+    most_genre = {}
+    # bulid a dict for genre count
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre in most_genre:
+            most_genre[genre] +=1
+        else:
+            most_genre[genre] =1
+    # use max to find the most one
+    return max(most_genre, key=most_genre.get)
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    friends_movies = set()
 
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    user_unique = []
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_movies:
+            user_unique.append(movie)
+    return user_unique
+
+def get_friends_unique_watched(user_data):
+    friends_movies = set()
+    
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    
+    user_watched = {movie["title"] for movie in user_data["watched"]}
+    
+    friends_unique = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_watched and movie not in friends_unique:
+                friends_unique.append(movie)
+    
+    return friends_unique
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
-
+def get_available_recs(user_data):
+    recommended_movies = []
+    if "subscriptions" in user_data and user_data["subscriptions"]:
+        for movie in  get_friends_unique_watched(user_data):
+            if movie["host"] in user_data["subscriptions"]:
+                recommended_movies.append(movie)

```

Comment:
Some other ways to approach filtering the `unique_watched` list could be using a list comprehension or using the `filter` function with a lambda function:

```python
# list comprehension
rec_movies = [movie for movie in unique_watched if movie["host"] in user_data["subscriptions"]]

# filter with a lambda function
rec_movies = filter(lambda movie: movie["host"] in user_data["subscriptions"], unique_watched)
```

(These lines are a bit long, in practice we might split a statement like these across lines, use some extra variables, or shorten some naming to keep under the PEP8 guide of 79 characters max per line.)

More info on Lambda functions in Python in case it’s useful: https://realpython.com/python-lambda/